[Inductor/Triton] Upcast FP16/BF16 math reductions to FP32

arui-meta opened the pull request:
Summary:
Triton compiler does not automatically promote fp16/bf16 reductions to fp32  accumulation. This will result in significant accuracy issue.

This diff will upcast the input to FP32 for all math reductions `["welford_reduce", "welford_combine", "prod", "sum", "xor_sum"]`

Test Plan:
CI
```
python test/inductor/test_torchinductor.py TritonCodeGenTests.test_low_precision_reduction
```

Differential Revision: D65965032




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

arui-meta commented:
@pytorchbot label "topic: not user facing"

eellison reviewed (commented):
clearing review queue - being reviewed internally

blaine-rister commented:
Left a review internally.

arui-meta committed:
[Inductor/Triton] Upcast FP16/BF16 math reductions to FP32 (#141052)

Summary:

Triton compiler does not automatically promote fp16/bf16 reductions to fp32  accumulation. This will result in significant accuracy issue.

This diff will upcast the input to FP32 for all reductions

Test Plan:
CI
```
python test/inductor/test_torchinductor.py TritonCodeGenTests.test_low_precision_reduction
```

Differential Revision: D65965032

blaine-rister reviewed (approved):
Nice work! LGTM.
